### PR TITLE
Comments: Jetpack comments tree workaround

### PR DESCRIPTION
**Pull request by Copons**
Introduce a new `QuerySiteCommentsList` component, as both a workaround to reliably fetch a list of comment IDs from Jetpack's remote sites, and for general use by anyone who might need to call this endpoint: https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/comments/.

`QuerySiteCommentsList` accepts as props all the query parameters listed in the [endpoint docs](https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/comments/), making it extensible enough for general use, and not limited for this workaround.
The `fetchCommentsList` data layer action, then takes care of building the actual query object, omitting the unnecessary props, and setting defaults for missing parameters.

---

The workaround is required in Comment Management because using the established "comments tree" method against the shadow site is currently not reliable because of sync issues on deleted comments, and there is no other way to retrieve the total count of comments of a given status and a given type.

This alternative approach, enabled only for Jetpack sites, consists in querying the `/sites/$site/comments/` endpoint for the maximum number of comments allowed (100), and immediately populate both the comments tree and the traditional comments list (`comments.items.{siteId)-{postId}`).

This means that at first load, Jetpack sites will have 5 pages top (considering `COMMENTS_PER_PAGE = 20`).

When changing a page, we dispatch a new request, this time offset by `( page - 1 ) * COMMENTS_PER_PAGE`, in order to load, at least 1 and at most 4 new pages of comments.

cc @Automattic/lannister

### CODE REVIEW

File: client/state/data-layer/wpcom/sites/comments/comment-tree/index.js [hunk 1]
```diff
@@ -67,8 +69,62 @@ const announceFailure = ( { dispatch, getState }, { query } ) => {
...
+		siteId,
+		status,
+		type: 'comment',
+	};
+
+	dispatch(
+		http(
+			{
+				method: 'GET',
+				path: `/sites/${ siteId }/comments`,
```
Comment (gwwar) [inline client/state/data-layer/wpcom/sites/comments/comment-tree/index.js]: @Copons we already have this action under `fetchCommentsList`. Note that we should follow folder conventions noted here: https://github.com/Automattic/wp-calypso/tree/master/client/state/data-layer#file-structure

Since this is an api abstraction that other folks might use, the original name is a better fit.

File: client/state/comments/actions.js [hunk 2]
```diff
@@ -17,6 +17,7 @@ import {
 	COMMENTS_WRITE,
 	COMMENT_REQUEST,
 	COMMENTS_TREE_SITE_REQUEST,
+	COMMENTS_TREE_JETPACK_SITE_REQUEST,
```
Comment (gwwar) [inline client/state/comments/actions.js]: No need for the new request type or action. This is available as `COMMENTS_LIST_REQUEST` and `requestCommentsList()` respectively

File: client/components/data/query-site-comments-tree/query-jetpack-site-comments-tree.jsx [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ */
+import { Component, PropTypes } from 'react';
+import { connect } from 'react-redux';
+
+/**
+ * Internal dependencies
+ */
+import { requestCommentsTreeForJetpackSite } from 'state/comments/actions';
+
+export class QueryJetpackSiteCommentsTree extends Component {
```
Comment (gwwar) [inline client/components/data/query-site-comments-tree/query-jetpack-site-comments-tree.jsx]: Let's rename this to `<QueryCommentsList />` so it's more clear. We'll be confused later if the comment-tree endpoint is available in Jetpack

File: client/state/data-layer/wpcom/sites/comments/comment-tree/index.js [hunk 4]
```diff
@@ -67,8 +69,62 @@ const announceFailure = ( { dispatch, getState }, { query } ) => {
 	dispatch( errorNotice( error ) );
 };
 
+const fetchCommentsTreeForJetpackSites = ( { dispatch }, action ) => {
+	const { page, siteId, status } = action.query;
+	const query = {
+		fields: 'ID,parent,post,status',
+		number: 100,
+		offset: ( page - 1 ) * 20, // see CommentList COMMENTS_PER_PAGE constant
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/comments/comment-tree/index.js]: note that there is no `page` stored in the database.

File: client/components/data/query-site-comments-tree/query-jetpack-site-comments-tree.jsx [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+		if ( pageChange || statusChange || siteId !== this.props.siteId ) {
+			this.request( ( pageChange && ! statusChange ) ? 'add' : 'replace' );
+		}
+	}
+
+	request( strategy = 'replace' ) {
+		const { page, siteId, status } = this.props;
+		if ( siteId ) {
+			this.props.requestCommentsTreeForJetpackSite( { page, siteId, status, strategy } );
```
Comment (gwwar) [inline client/components/data/query-site-comments-tree/query-jetpack-site-comments-tree.jsx]: Can we pull `strategy` out of the query arg. I 😆 tried to look it up in documentation https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/comments/

**Comment by gwwar** [comment]
Nice @Copons Overall this works great with the limitations we have. ✨  A few things we probably want before 🚢 
- tidy up the query component/data-layer pieces as noted in comments
- populate not only the tree but also comments item, which I think should happen naturally if you use the existing data-layer hooks.

File: client/my-sites/comments/comment-list/index.jsx [hunk 6]
```diff
@@ -379,11 +379,14 @@ export class CommentList extends Component {
...
-				{ ! isJetpack &&
-					<QuerySiteCommentsTree siteId={ siteId } status={ status } />
+				{ isJetpack
+					? <QuerySiteCommentsList
+						number={ 100 }
+						offset={ ( page - 1 ) * COMMENTS_PER_PAGE }
+						siteId={ siteId }
+						status={ status }
+					/>
+					: <QuerySiteCommentsTree siteId={ siteId } status={ status } />
```
Comment (dmsnell) [inline client/my-sites/comments/comment-list/index.jsx]: sometimes when the formatting like this is bad I just ditch the ternary in favor of two separate conditionals; your choice

```js
{ isJetpack && (
	<QuerySiteCommentsList
		number={ 100 }
		…
	/>
) }
{ ! isJetpack && (
	<QuerySiteCommentsTree … />
) }
```

File: client/state/comments/actions.js [hunk 7]
```diff
@@ -56,6 +56,17 @@ export function requestPostComments( {
...
 
+/**
+ * Creates an action that request a list of comments for a given query.
+ * Except the two query properties descibed here, this function accepts all query parameters
+ * listed in the API docs:
+ * @see https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/comments/
+ *
+ * @param {Object} query API call parameters
+ * @param {String} query.listType Type of list to return (required as 'site')
+ * @param {Number} query.siteId Site identifier
```
Comment (dmsnell) [inline client/state/comments/actions.js]: nice JSDoc explanations!

File: client/state/data-layer/wpcom/sites/comments/index.js [hunk 8]
```diff
@@ -138,31 +139,43 @@ export const receiveCommentError = ( { dispatch, getState }, { siteId, commentId
 
 // @see https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/comments/
 export const fetchCommentsList = ( { dispatch }, action ) => {
-	const { query } = action;
-
-	if ( ! query || 'site' !== query.listType ) {
+	if ( ! action.query || 'site' !== action.query.listType ) {
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/comments/index.js]: should probably leave a note where why we're aborting if the query is provided; a link to where it's handled would also be great with a `@see` tag

File: client/state/data-layer/wpcom/sites/comments/index.js [hunk 9]
```diff
@@ -138,31 +139,43 @@ export const receiveCommentError = ( { dispatch, getState }, { siteId, commentId
...
+				query,
 			},
 			action,
 		),
 	);
 };
 
-export const addComments = ( { dispatch }, { query: { siteId } }, next, { comments } ) => {
+export const addComments = ( { dispatch }, { query: { siteId, status } }, next, { comments } ) => {
+	// Initialize the comments tree
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/comments/index.js]: Might actually be more accurate to say something like "Make sure comments tree is initialized"

actually, why do we need to do this if we are also listening for these actions and storing a separate `isInitialized` state value?

File: client/state/data-layer/wpcom/sites/comments/test/index.js [hunk 10]
```diff
@@ -29,10 +29,10 @@ describe( '#addComments', () => {
 
 	beforeEach( () => ( dispatch = spy() ) );
 
-	it( 'should dispatch no actions for no comments', () => {
+	it( 'should dispatch one action for no comments', () => {
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/comments/test/index.js]: maybe this isn't so descriptive. why the change form none to one?

how about instead we test for a positive _and_ descriptive behavior, like `it( 'should dispatch a tree initialization action' )` (of course, pending the value of that action being dispatched)

File: client/state/data-layer/wpcom/sites/comments/index.js (hunk 8 above)
Comment (Copons) [inline client/state/data-layer/wpcom/sites/comments/index.js]: Do you mean why we abort if `query` is provided but `query.listType !== 'site'`?
Or in other words, why do we even need `listType`, if we expect it to be one specific value, but then we don't use it anywhere?

I honestly have no idea.
Probably the idea here is to have this extendable to fetching either site or post comments, but I'm not sure.

File: client/state/data-layer/wpcom/sites/comments/index.js (hunk 9 above)
Comment (Copons) [inline client/state/data-layer/wpcom/sites/comments/index.js]: For the whole `isLoading` / `emptyContent` handling of Comment Management lists, we set `comments.treesInitialized` (so that we know that an empty tree is actually empty and not loading) by listening to `COMMENTS_TREE_SITE_ADD`, which normally is dispatched on `fetchTree` success.

In case of JP sites, we never call `fetchTree`, but we populate the trees as a side effect of `fetchCommentsList`, listening to `COMMENTS_RECEIVE`.

If we receive even a single comment, we don't need to set `treesInitialized`, because a non-empty tree is good enough for the various tree selectors.

But if we receive an empty response, we need to "manually" initialize the tree, so that the tree selectors actually know that is loaded and empty.

File: client/my-sites/comments/comment-list/index.jsx (hunk 6 above)
Comment (Copons) [inline client/my-sites/comments/comment-list/index.jsx]: Fun fact: it was in fact separate ifs before, but I replaced it with a ternary because I was expecting your customary "you can use a ternary here" comment. 😆

File: client/state/data-layer/wpcom/sites/comments/index.js (hunk 8 above)
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/comments/index.js]: ah I see now - we're not adding a condition, just unwrapping `action`

File: client/state/data-layer/wpcom/sites/comments/index.js (hunk 9 above)
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/comments/index.js]: Sounds like a great comment!

File: client/state/data-layer/wpcom/sites/comments/index.js (hunk 8 above)
Comment (gwwar) [inline client/state/data-layer/wpcom/sites/comments/index.js]: Maybe cleaner to do:
```jsx
if( get( action, 'query.listType' ) !== 'site' ) {
```

Comment (Copons) [inline client/state/data-layer/wpcom/sites/comments/index.js]: Although, I'd really like to know what would be `listType` use case, as AFAIK it's not used by anyone anywhere...
